### PR TITLE
[SYCL][FPGA] Emit multiple IR variants for the IVDep attribute

### PR DESCRIPTION
**Pull request by AGindinson**
By emitting the legacy variant of the LLVM IR alongside the newer
representation of the attribute, backwards compatibility with any
existing BE implementation. A smooth transition period is thus
achieved for the aforementiond BE - until it's able to consume the
new LLVM IR, it has an option to simply ignore the unknown metadata.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

### CODE REVIEW

File: clang/lib/CodeGen/CGLoopInfo.cpp [hunk 1]
```diff
@@ -476,6 +476,23 @@ EmitIVDepLoopMetadata(LLVMContext &Ctx,
   LoopProperties.push_back(MDNode::get(Ctx, MD));
 }
 
+/// Setting the legacy LLVM IR representation of the ivdep attribute
```
Comment (erichkeane) [inline clang/lib/CodeGen/CGLoopInfo.cpp]:
```suggestion
/// Setting the legacy LLVM IR representation of the ivdep attribute.
```

File: clang/lib/CodeGen/CGLoopInfo.cpp [hunk 2]
```diff
@@ -476,6 +476,23 @@ EmitIVDepLoopMetadata(LLVMContext &Ctx,
...
+static void EmitLegacyIVDepLoopMetadata(
+    LLVMContext &Ctx, llvm::SmallVectorImpl<llvm::Metadata *> &LoopProperties,
+    const LoopAttributes::SYCLIVDepInfo &I) {
+  // Emit "enable" metadata upon the infinite safelen specification
```
Comment (erichkeane) [inline clang/lib/CodeGen/CGLoopInfo.cpp]: Can you reword this one?  The 'upon the infinite safelen specification' doesn't really parse for me.

File: clang/test/CodeGenSYCL/intel-fpga-ivdep-array.cpp [hunk 3]
```diff
@@ -186,28 +186,32 @@ int main() {
 // CHECK-DAG: ![[IVDEP_MUL_ARR_VAL]] = !{!"llvm.loop.parallel_access_indices", ![[IDX_GROUP_A_MUL_ARR]], ![[IDX_GROUP_B_MUL_ARR]], i32 5}
 // CHECK-DAG: ![[IVDEP_MUL_ARR_INF]] = !{!"llvm.loop.parallel_access_indices", ![[IDX_GROUP_C_MUL_ARR]], ![[IDX_GROUP_D_MUL_ARR]]}
 
+// Find the single instance of a legacy "IVDep enable" MD node
```
Comment (erichkeane) [inline clang/test/CodeGenSYCL/intel-fpga-ivdep-array.cpp]:
```suggestion
// Find the single instance of a legacy "IVDep enable" MD node.
```

File: clang/test/CodeGenSYCL/intel-fpga-ivdep-embedded-loops.cpp [hunk 4]
```diff
@@ -164,18 +164,22 @@ int main() {
   return 0;
 }
 
+// Find the single instance of a legacy "IVDep enable" MD node
```
Comment (erichkeane) [inline clang/test/CodeGenSYCL/intel-fpga-ivdep-embedded-loops.cpp]:
```suggestion
// Find the single instance of a legacy "IVDep enable" MD node.
```

File: clang/test/CodeGenSYCL/intel-fpga-ivdep-global.cpp [hunk 5]
```diff
@@ -94,31 +94,35 @@ int main() {
   return 0;
 }
 
+// Find recurring instances of legacy "IVDep enable/safelen" MD nodes
```
Comment (erichkeane) [inline clang/test/CodeGenSYCL/intel-fpga-ivdep-global.cpp]:
```suggestion
// Find recurring instances of legacy "IVDep enable/safelen" MD nodes.
```

File: clang/lib/CodeGen/CGLoopInfo.cpp (hunk 1 above)
Comment (AGindinson) [reply clang/lib/CodeGen/CGLoopInfo.cpp]: Done along with other comments.

File: clang/lib/CodeGen/CGLoopInfo.cpp (hunk 2 above)
Comment (AGindinson) [reply clang/lib/CodeGen/CGLoopInfo.cpp]: Done - please check if the comment makes sense now.

**Comment by AGindinson** [comment]
Updated the main commit message.

**Comment by AGindinson** [comment]
> So, we are OK to have two decorations on SPIR-V level?

Yep, the Translator retains support for both IR variants, translating these into varying SPIR-V constructs.

**Pull request merged**